### PR TITLE
Jj output ctf

### PR DESCRIPTION
**Pull request by JorMaister**

### CODE REVIEW

File: aretomo/convert/dataimport.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+import numpy as np
+import logging
+
+
+logger = logging.getLogger(__name__)
+
+import pyworkflow.utils as pwutils
+from pwem.objects import CTFModel
+
+with pwutils.weakImport('tomo'):
```
Comment (azazellochg) [inline aretomo/convert/dataimport.py:38]: do we really need the weakImport for plugins that are designed for tomo only?

File: aretomo/protocols/protocol_aretomo.py [hunk 2]
```diff
@@ -631,6 +739,18 @@ def getOutputSetOfTiltSeries(self,
                                        outputSetOfTiltSeries)
         return getattr(self, outputName)
 
+    def getOutputSetOfCtfs(self) -> SetOfCTFTomoSeries:
+        outputCtfs = getattr(self, OUT_CTFS, None)
+        if outputCtfs:
+            outputCtfs.enableAppend()
+        else:
+            outputCtfs = SetOfCTFTomoSeries.create(self._getPath(), template='CTFmodels%s.sqlite')
+            outputCtfs.setSetOfTiltSeries(self.getOutputSetOfTiltSeries(OUT_TS))
```
Comment (azazellochg) [inline aretomo/protocols/protocol_aretomo.py:748]: what is the difference between newCTFTomoSeries.setTiltSeries(newTs) and outputCtfs.setSetOfTiltSeries(..)?

Comment (azazellochg) [reply aretomo/protocols/protocol_aretomo.py:748]: I got it, the outputCtfs is the set object.. :) Maybe we should rename methods to match..

**Comment by azazellochg** [comment]
Thank you very much, @JorMaister that's a lot of work! I will test this when I can

**Comment by JorMaister** [comment]
I'm changing the state to draft until I add the psd file and make the tests check for it and also a CUDA_LIB variable (I forgot that couple of things). I expect them to be tomorrow in the morning.

**Comment by azazellochg** [comment]
No worries, I'll try to review it this weekend.

**Comment by azazellochg** [comment]
I already have problems with the first test: align&rec with ctf, 3 threads, 2 TS, 1 GPU. The first aretomo process launches and somewhere in the middle of the run the second TS is started on the same GPU. This slows down processing tremendously. I cannot reduce the number of threads <3.

File: aretomo/protocols/protocol_aretomo.py [hunk 3]
```diff
@@ -522,6 +626,37 @@ def createOutputStep(self, tsObjId: int, tsId: str, tsFn: str):
...
             self._store(outputSetOfTiltSeries)
 
+            # Output set of CTF tomo series ----------------------------------------
+            if self.doEstimateCtf.get():
+                outputCtfs = self.getOutputSetOfCtfs()
+
+                newCTFTomoSeries = CTFTomoSeries()
+                newCTFTomoSeries.copyInfo(newTs)
+                newCTFTomoSeries.setTiltSeries(newTs)
+                newCTFTomoSeries.setObjId(ts.getObjId())
```
Comment (azazellochg) [inline aretomo/protocols/protocol_aretomo.py]: why do we need to set objId?

File: aretomo/protocols/protocol_aretomo.py [hunk 4]
```diff
@@ -522,25 +612,66 @@ def createOutputStep(self, tsObjId: int, tsId: str, tsFn: str):
...
-                           f"{self.outputSetOfTiltSeries.getSize()}")
+                           f"{getattr(self, OUT_TS).getSize()}")
         else:
             summary.append("Output is not ready yet.")
 
-        if self._saveInterpolated() and not self.makeTomo:
-            summary.append("*Interpolated TS stack may have a few "
-                           "dark tilt images removed.*")
+        if self.isFinished() and self.noneGeneratedMsg.get():
+            summary.append('*ERROR!*\n' + self.noneGeneratedMsg.get())
```
Comment (azazellochg) [inline aretomo/protocols/protocol_aretomo.py:663]: I'm still getting this error msg printed even when all outputs are produced correctly.

**Review by azazellochg — CHANGES_REQUESTED** [review]

**Review by JorMaister — COMMENTED** [review]
@azazellochg. why did you remove the comment regarding the CTF estimation? I think it's not clear why the behaviour relies on the Cs, so it should be explained somehow

**Comment by azazellochg** [comment]
ok. If you can make a shorted comment :) CTF est requires Cs, the rest is
required for dose-weighting
Best regards,
Grigory

--------------------------------------------------------------------------------
Grigory Sharov, Ph.D.

MRC Laboratory of Molecular Biology,
Francis Crick Avenue,
Cambridge Biomedical Campus,
Cambridge CB2 0QH, UK.
tel. +44 (0) 1223 267228 <+44%201223%20267228>
e-mail: ***@***.***

File: aretomo/protocols/protocol_aretomo.py (hunk 4 above)
Comment (JorMaister) [reply aretomo/protocols/protocol_aretomo.py:663]: Hi, @azazellochg. I'm not able to reproduce it. Could it be related to a missing git pull or something like that?

File: aretomo/protocols/protocol_aretomo.py (hunk 3 above)
Comment (JorMaister) [reply aretomo/protocols/protocol_aretomo.py]: Hi, @azazellochg,
You're right

File: aretomo/protocols/protocol_aretomo.py (hunk 4 above)
Comment (azazellochg) [reply aretomo/protocols/protocol_aretomo.py:667]: Okay, let me try again this evening

Comment (azazellochg) [reply aretomo/protocols/protocol_aretomo.py:667]: Indeed I'm not getting it. Now I realize that this code will never be executed.
```
self.noneGeneratedMsg.set(msg)
raise Exception(msg)
```
will raise an exception and fail the protocol, so the status will never be finished

Comment (JorMaister) [reply aretomo/protocols/protocol_aretomo.py:667]: Thanks, @azazellochg. The Exception is the desired behaviour as no protocol should finish its execution in green with a correct status if no output is generated (when there are outputs expected to be generated, of course). You are right regarding the summary message, it's dead code from a previous attempt. Let me delete those lines.

Comment (JorMaister) [reply aretomo/protocols/protocol_aretomo.py]: On the other hand, are you getting the message no matter how the execution goes, as said previously? I'm not getting it. This is what I get from the tests.

image

Or this if some fail, but not all:

image

Comment (azazellochg) [reply aretomo/protocols/protocol_aretomo.py]: Nope, it's all correct. I get the same as you now.

**Comment by JorMaister** [comment]
> somewhere in the middle of the run the second TS is started on the same GPU. This slows down processing tremendously. I cannot reduce the number of threads <3.

Hi, @azazellochg. After the last changes in pyworkflow, it's now solved. I'll change the default number of threads to 2, as the default value for the GPU is one GPU, thus there will be 1 thread for the step generator and the other for running the steps.

**Review by azazellochg — APPROVED** [review]
